### PR TITLE
Binding hash and cleanup

### PR DESCRIPTION
**Pull request by zeptognome**
Adds binding to hash to prevent mixing different bindings in a virtual stack
Cleanup of binding code as it never returns nil
Bound items always correctly default to soulbound now
bindType override from itemlink pulled into items instead of binding

General refactor of data.itemInfo not in scope of this PR

### CODE REVIEW

File: data/binding.lua [hunk 1]
```diff
@@ -81,9 +60,7 @@ function binding.GetItemBinding(itemID, itemLocation, itemLink)
       bindinginfo.binding = const.BINDING_SCOPE.QUEST
     end
   end -- isBound
-  if bindinginfo.binding == const.BINDING_SCOPE.UNKNOWN then
```
Comment (zeptognome) [inline data/binding.lua:84]: worst case of no matches the defaults of unknown/bound=false falls thru

File: data/search.lua [hunk 2]
```diff
@@ -201,7 +201,7 @@ function search:Add(item)
     search:addStringToIndex(self.indicies.equipmentlocation, _G[item.itemInfo.itemEquipLoc], item.slotkey)
   end
 
-  if item.bindingInfo ~= nil then
+  if const.BINDING_MAP[item.bindingInfo.binding] ~= "" then
```
Comment (zeptognome) [inline data/search.lua:204]: not needed if indexing empty strings is safe

File: data/items.lua [hunk 3]
```diff
@@ -799,6 +800,18 @@ function items:GetCategory(data)
   return category
 end
 
+---@param itemLink string
+---@return Enum.ItemBind?
+function items:GetBindTypeFromLink(itemLink)
+  -- itemLink has better information for items, but no information for pet or keystone links
+  local bindType = nil
+  if ( strfind(itemLink, "item:") ) then
+    bindType, _, _, _ = select(14,C_Item.GetItemInfo(itemLink ))
```
Comment (zeptognome) [inline data/items.lua:809]: this needs a bigger refactor to genericly handle non-'item:' links with a fallback to itemID for pets and keys except for any of itemInfo's properties that aren't going to be replace by C_Item.foo APIs

I see this eventually becoming a ---@return itemInfo function

File: data/items.lua [hunk 4]
```diff
@@ -823,6 +836,7 @@ function items:AttachItemInfo(data, kind)
   itemStackCount, itemEquipLoc, itemTexture,
   sellPrice, classID, subclassID, bindType, expacID,
   setID, isCraftingReagent = C_Item.GetItemInfo(itemID)
+  bindType = self:GetBindTypeFromLink(itemLink) or bindType  --link overrides itemID if set
```
Comment (zeptognome) [inline data/items.lua:839]: placeholder override until refactor above

File: data/items.lua (hunk 3 above)
Comment (juntereiner) [reply data/items.lua:809]: you can just do
```
local bindType, _, _, _ = select(14,C_Item.GetItemInfo(itemLink ))
if bindType == nil then
    bindType, _, _, _ = select(14,C_Item.GetItemInfo(itemID ))
end
```

Comment (zeptognome) [reply data/items.lua:809]: > you can just do
I can actually just do
bindType, _, _, _ = select(14,C_Item.GetItemInfo(itemLink ))

as the main AttachItemInfo now handles a falsy result,
but as I noted, this is a placeholder for a bigger rework

File: data/items.lua [hunk 5]
```diff
@@ -799,6 +800,17 @@ function items:GetCategory(data)
   return category
 end
 
+---@param itemLink string
+---@return Enum.ItemBind?
+function items:GetBindTypeFromLink(itemLink)
+  -- itemLink has better information for items, but no information for pet or keystone links
+  local bindType = nil
+  if ( strfind(itemLink, "item:") ) then
```
Comment (Cidan) [inline data/items.lua:808]: fix spacing here -- no space between parathesis

**Review by Cidan — APPROVED** [review]

**Pull request merged**